### DISCUSSION TITLE
Chunk as a Non-Empty data structure

### DISCUSSION THREAD

**Discussion by akilegaspi** [typelevel/fs2#3440] [Ideas]
I think it makes sense for the idea of Chunk to be Non-Empty since generally the idea of a Chunk is a division of _something_ whereas a Chunk should only exist if there is something that has been divided and a subset of that something is the Chunk.

I'm not sure what is the implication in current Fs2 underlying mechanism

**Comment by mpilquist** [comment]
We used to have `NonEmptyChunk` as a subtype of `Chunk` (all the way back in version 0.9, introduced in https://github.com/typelevel/fs2/pull/718). It was subsequently removed during the development of the 0.10 release, due in part due to clunkiness in usage.

I don't think we ever really considered instead removing the ability to have empty chunks entirely. Doing so wouldn't require deep changes to the core interpreter but it would certainly impact lots of derived operations and user code. For example, `filter` on `Stream` is implemented as `mapChunks(_.filter(p))`. What type would `Chunk#filter` return if chunks can't be empty? Every constructor from a collection to a chunk would need to handle the empty case and signal it, which users would then need to handle -- e.g., `Chunk.from(collection)` would need to return `Option[Chunk[A]]` instead of `Chunk[A]`. There'd be a bunch of annoying issues like this to work through and the end state would like be more annoying to work with than what we have now.

**Comment by akilegaspi** [comment]
Thank you for the clear answer
